### PR TITLE
No "weird" floats in const fn {from,to}_bits

### PR DESCRIPTION
**Pull request by workingjubilee**
I suspect this code is subtly incorrect and that we don't even e.g. use x87-style floats in CTFE, so I don't have to guard against that case. A future PR will be hopefully removing them from concern entirely, anyways. But at the moment I wanted to get this rolling because small questions like that one seem best answered by review.

r? @oli-obk 
cc @eddyb @thomcc

### CODE REVIEW

**Comment by oli-obk** [comment]
oof. What you write makes sense, but I have no way of verifying it beyond looking at the CTFE logic and agreeing that it makes sense.

> we don't even e.g. use x87-style floats in CTFE, so I don't have to guard against that case.

yea, we don't. Our impl is "perfectly IEEE compliant". I am fully aware how useless that fact is, considering hardware and LLVM do what they do.

File: library/core/src/num/f32.rs [hunk 1]
```diff
@@ -568,15 +572,70 @@ impl f32 {
...
+            // Thus, a value may compare unequal to infinity, despite having a "full" exponent mask.
+            FpCategory::Infinite
+        } else if self.is_nan() {
+            // And it may not be NaN, as it can simply be an "overextended" finite value.
+            FpCategory::Nan
+        } else {
+            // However, std can't simply compare to zero to check for zero, either,
+            // as correctness requires avoiding equality tests that may be Subnormal == -0.0
+            // because it may be wrong under "denormals are zero" and "flush to zero" modes.
+            // Most of std's targets don't use those, but they are used for thumbv7neon".
```
Comment (eternaleye) [inline library/core/src/num/f32.rs:597]: Missing double quote?

File: library/core/src/num/f64.rs [hunk 2]
```diff
@@ -567,15 +573,67 @@ impl f64 {
...
+        //
+        // Thus, a value may compare unequal to infinity, despite having a "full" exponent mask.
+        // And it may not be NaN, as it can simply be an "overextended" finite value.
+        if self.is_nan() {
+            FpCategory::Nan
+        } else {
+            // However, std can't simply compare to zero to check for zero, either,
+            // as correctness requires avoiding equality tests that may be Subnormal == -0.0
+            // because it may be wrong under "denormals are zero" and "flush to zero" modes.
+            // Most of std's targets don't use those, but they are used for thumbv7neon".
```
Comment (eternaleye) [inline library/core/src/num/f64.rs:596]: Missing double quote?

**Comment by workingjubilee** [comment]
> oof. What you write makes sense, but I have no way of verifying it beyond looking at the CTFE logic and agreeing that it makes sense.
> 
> > we don't even e.g. use x87-style floats in CTFE, so I don't have to guard against that case.
> 
> yea, we don't. Our impl is "perfectly IEEE compliant". I am fully aware how useless that fact is, considering hardware and LLVM do what they do.

That's fine! It's my job to bring those two more into alignment, one way or another. :smiling_imp:

**Comment by workingjubilee** [comment]
Fix comments and removed unnecessary x87-specific logic for const time.

**Comment by oli-obk** [comment]
This gets us closer to floats in const fn. I'll address the use of const_eval_select after this is merged.

@bors r+

**Pull request merged**